core/src/main/java/com/linecorp/armeria/internal/client/dns/SearchDomainDnsResolver.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.client.dns;
+
+import static com.spotify.futures.CompletableFutures.exceptionallyCompletedFuture;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.spotify.futures.CompletableFutures;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.AbstractUnwrappable;
+
+import io.netty.handler.codec.dns.DnsQuestion;
+import io.netty.handler.codec.dns.DnsRecord;
+
+final class SearchDomainDnsResolver extends AbstractUnwrappable<DnsResolver> implements DnsResolver {
+
+    private final List<String> searchDomains;
+    private final int ndots;
+    private volatile boolean closed;
+
+    SearchDomainDnsResolver(DnsResolver delegate, List<String> searchDomains, int ndots) {
+        super(delegate);
+        this.searchDomains = searchDomains;
+        this.ndots = ndots;
+    }
+
+    @Override
+    public CompletableFuture<List<DnsRecord>> resolve(DnsQuestionContext ctx, DnsQuestion question) {
+        final SearchDomainQuestionContext searchDomainCtx =
+                new SearchDomainQuestionContext(question, searchDomains, ndots);
+        final DnsQuestion firstQuestion = searchDomainCtx.nextQuestion();
+        assert firstQuestion != null;
+        return resolve0(ctx, searchDomainCtx, firstQuestion);
+    }
+
+    private CompletableFuture<List<DnsRecord>> resolve0(DnsQuestionContext ctx,
+                                                        SearchDomainQuestionContext searchDomainCtx,
+                                                        DnsQuestion question) {
+        if (closed) {
+            return exceptionallyCompletedFuture(new IllegalStateException("resolver is closed already"));
```

Comment:
Ditto - how about blocklisting the usage of `CompletableFutures.exceptionallyCompletedFuture()` using a Checkstyle rule?